sdk/core/azure_core_amqp/src/connection.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved.
+// cspell: words amqp sasl
+
+use super::value::{AmqpOrderedMap, AmqpSymbol, AmqpValue};
+use azure_core::error::Result;
+use std::fmt::Debug;
+use time::Duration;
+use url::Url;
+
+pub struct AmqpConnectionOptions {
+    pub(crate) max_frame_size: Option<u32>,
+    pub(crate) channel_max: Option<u16>,
+    pub(crate) idle_timeout: Option<Duration>,
+    pub(crate) outgoing_locales: Option<Vec<String>>,
+    pub(crate) incoming_locales: Option<Vec<String>>,
+    pub(crate) offered_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) desired_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) properties: Option<AmqpOrderedMap<AmqpSymbol, AmqpValue>>,
+    pub(crate) buffer_size: Option<usize>,
+}
+
+impl AmqpConnectionOptions {
+    pub fn builder() -> builders::AmqpConnectionOptionsBuilder {
+        builders::AmqpConnectionOptionsBuilder::new()
+    }
+}
+
+pub trait AmqpConnectionTrait {
+    #[allow(unused_variables)]
+    fn open(
+        &self,
+        name: impl Into<String>,
+        url: Url,
+        options: Option<AmqpConnectionOptions>,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        async {
+            unimplemented!();
```

Comment:
Was this necessary? You typically just declare the function without a body if it must be implemented. You only have to do this when returning an unsized thing like `Self` since, in a trait, `Self` doesn't implemented `Sized`. rust-analyzer and VSCode will easily add all "abstract" functions then, too. It won't this way.

Reply:
This was a suggestion from the rustacians teams channel - they suggested providing a default implementation on these trait methods so that they don't need to be implemented in every new trait implementation.

I'm not wedded to this pattern at all.

Reply:
Oh, so these are *required* implementations? Than why err at all? Just to reinforce develop intent?